File: SRChoco/SRWindowManager/SRWindowManager.m
//
//  SRWindowManager.m
//  SRChocoDemo-OSX
//
//  Created by Heeseung Seo on 2014. 12. 22..
//  Copyright (c) 2014 Seorenn. All rights reserved.
//

#import "SRWindowManager.h"

@implementation SRWindowManager

@synthesize detecting = _detecting;

+ (SRWindowManager *)sharedManager {
    static SRWindowManager *instance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [[SRWindowManager alloc] init];
    });
    return instance;
}

- (id)init {
    self = [super init];
    if (self) {
        NSNotificationCenter *nc = [[NSWorkspace sharedWorkspace] notificationCenter];
        [nc addObserver:self
               selector:@selector(didActivateWindowNotification:)
                   name:NSWorkspaceDidActivateApplicationNotification
                 object:nil];
    }
    return self;
}

- (void)dealloc {
    [[[NSWorkspace sharedWorkspace] notificationCenter] removeObserver:self];
}

- (void)startDetect {
    _detecting = YES;
}

- (void)stopDetect {
    _detecting = NO;
}

- (void)didActivateWindowNotification:(NSNotification *)notification {
    if (_detecting == NO || self.delegate == nil || [self.delegate respondsToSelector:@selector(windowManager:detectWindowActivation:)] == NO) return;
    
    NSRunningApplication *app = [notification.userInfo objectForKey:NSWorkspaceApplicationKey];
    [self.delegate windowManager:self detectWindowActivation:app];
}

- (__autoreleasing NSArray *)processes {
    NSMutableArray *results = [[NSMutableArray alloc] init];
    NSArray *apps = [[NSWorkspace sharedWorkspace] runningApplications];
    for (NSRunningApplication *app in apps) {
        SRApplicationWindow *window = [[SRApplicationWindow alloc] initWithRunningApplication:app];
        if (!window) continue;
        
        [results addObject:window];
    }
    
    return results;
}

- (__autoreleasing NSArray *)windows {
    NSMutableArray *results = [[NSMutableArray alloc] init];
    CFArrayRef list = CGWindowListCopyWindowInfo(kCGWindowListOptionOnScreenOnly | kCGWindowListExcludeDesktopElements, kCGNullWindowID);
    if (!list) return nil;
    
    NSInteger count = CFArrayGetCount(list);
    for (int i=0; i < count; i++) {
        CFDictionaryRef info = CFArrayGetValueAtIndex(list, i);
        if (!info) continue;
        
        CFNumberRef pidNumber = CFDictionaryGetValue(info, kCGWindowOwnerPID);
        pid_t pid;
        CFNumberGetValue(pidNumber, kCFNumberIntType, &pid);
        
        SRApplicationWindow *window = [[SRApplicationWindow alloc] initWithPID:pid];
        if (!window) continue;
            
        [results addObject:window];
    }
    
    CFRelease(list);
    
    return results;
}

- (__autoreleasing NSArray *)applicationWindows {
    NSMutableArray *results = [[NSMutableArray alloc] init];

    NSArray *apps = [[NSWorkspace sharedWorkspace] runningApplications];
    for (NSRunningApplication *app in apps) {
        if (app.activationPolicy != NSApplicationActivationPolicyRegular) continue;
        
        SRApplicationWindow *window = [[SRApplicationWindow alloc] initWithRunningApplication:app];
        [results addObject:window];
    }

    return results;
}

@end

File: SRChoco/SRWindowManager/SRWindowManager.swift
//
// SRWindowManager.swift
// SRChoco
//
// Created by Seorenn.
// Copyright (c) 2014 Seorenn. All rights reserved.
//

#if os(OSX)

import Cocoa
    
class SRWindowManager {
    // singleton instance type
    struct StaticInstance {
        static var dispatchToken: dispatch_once_t = 0
        static var instance:SRWindowManager?
    }
    
    // singleton factory
    class func sharedManager() -> SRWindowManager {
        dispatch_once(&StaticInstance.dispatchToken) {
            StaticInstance.instance = SRWindowManager()
        }
        return StaticInstance.instance!
    }
    
    deinit {
        let nc = NSWorkspace.sharedWorkspace().notificationCenter
        nc.removeObserver(self)
    }
    
    class var processes: [NSRunningApplication] {
        var results = [NSRunningApplication]()
        let apps = NSWorkspace.sharedWorkspace().runningApplications as [NSRunningApplication]
        for app in apps {
            results.append(app)
        }
        return results
    }

    class var windowProcesses: [NSRunningApplication] {
        var apps = [NSRunningApplication]()
        let list = CGWindowListCopyWindowInfo(CGWindowListOption(kCGWindowListExcludeDesktopElements | kCGWindowListOptionOnScreenOnly), CGWindowID(0))
        let windowInfos = list.takeRetainedValue() as Array
        for info in windowInfos as Array<Dictionary<NSString, AnyObject>> {
            let pidPtr: AnyObject? = info[kCGWindowOwnerPID]
            let pidInt = pidPtr as Int
            let pid = CInt(pidInt)
            if let app = NSRunningApplication(processIdentifier: pid) {
                apps.append(app)
            }
        }
        return apps
    }
    
    func detectWindowChanging(block:((NSRunningApplication!) -> Void)!) {
        let nc = NSWorkspace.sharedWorkspace().notificationCenter
        nc.addObserverForName(NSWorkspaceDidActivateApplicationNotification, object:nil, queue:NSOperationQueue.mainQueue(), usingBlock: {(notification: NSNotification!) -> Void in
            let app = notification!.userInfo![NSWorkspaceApplicationKey] as NSRunningApplication!
            block(app)
        })
    }
}
    
#endif  //#if os(OSX)
